### PR TITLE
Integrate GKE operator

### PR DESCRIPTION
**Pull request by cmurphy**
Add GKE v2 integration controller and cluster refresher.

https://github.com/rancher/rancher/issues/31011
https://github.com/rancher/rancher/issues/31010
https://github.com/rancher/rancher/issues/31461
https://github.com/rancher/rancher/issues/31895
https://github.com/rancher/rancher/issues/31110

### CODE REVIEW

**Comment by cmurphy** [comment]
Marking this as ready for review so I can start working on feedback even though it is awaiting final approval of https://github.com/rancher/rancher/pull/31674

**Review by rmweir — CHANGES_REQUESTED** [review]
Skipped any parts related to syncing as that is subject to change. We should probably create a package contaaining logic that is used in multiple controllers such as deploy<provider>Operator, setUnknown, setTrue, setFalse, aand generateValuesYaml.

File: pkg/api/norman/customization/cluster/validator.go
Comment (cmurphy) [reply pkg/api/norman/customization/cluster/validator.go:565]: Why would this be different from the EKS credential? https://github.com/rancher/rancher/blob/1031efc0fcf9078037c841c4d8bcc3af92b78360/pkg/api/norman/customization/cluster/validator.go#L355-L360

I could move it down past line 582 to validate for creates only but I think the credential still needs to be present and validated in the import case?

File: pkg/api/norman/customization/cluster/validator.go
Comment (cmurphy) [reply pkg/api/norman/customization/cluster/validator.go:665]: It would end up getting rejected by the operator. Fixing here.

File: pkg/api/norman/customization/cluster/validator.go
Comment (cmurphy) [reply pkg/api/norman/customization/cluster/validator.go:691]: Yes. Fixing.

File: pkg/controllers/management/clusterdeploy/clusterdeploy.go
Comment (cmurphy) [reply pkg/controllers/management/clusterdeploy/clusterdeploy.go:129]: Thanks for bring this up because I was confused about it, see the commit message:

```
Create system cluster token on first sync attempt

If a cluster provisioner has deployed a remote cluster with a private
control plane endpoint, it will need to have the cluster agent deployed
manually. Without this patch, it will not be able to serve the agent
YAML manifest yet because doSync exits early since it cannot find any
nodes. In order to be able to generate the agent manifest URL, this
change goes ahead and generates the cluster registration token as soon
as the cluster is deployed.
```

In short, when deploying a private cluster in GKE I couldn't do the manual agent install step without this. I don't know why it's a problem here and not a problem for EKS, do you?

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:235]: Oops, fixed

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:268]: Copypasta from eks_cluster_handler. In this case `generateSATokenWithPublicAPI` never returns an error if mustTunnel is true but I'll fix it to future-proof it.

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cbron) [inline pkg/controllers/management/gke/gke_cluster_handler.go:214]: This is slight departure from EKS where we always check imported first. We are saying now that any type of cluster where upstream is nil we want to SetInitialUpstreamSpec

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cbron) [inline pkg/controllers/management/gke/gke_cluster_handler.go:241]: Shouldnt we be making sure there is at least one nodegroup here ?

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cbron) [inline pkg/controllers/management/gke/gke_cluster_handler.go:624]: Good thing we aren't using bearer

File: pkg/controllers/management/clusterdeploy/clusterdeploy.go
Comment (cbron) [reply pkg/controllers/management/clusterdeploy/clusterdeploy.go:129]: @thedadams can you review this, and also the last two commits in this PR for private cluster approach.

**Review by cbron — COMMENTED** [review]
It's in description but this depends on integration PR first. 
Kinda sad we have to copy so much code and not use shared package but oh well.

File: go.mod
Comment (cmurphy) [reply go.mod:13]: `go mod tidy` fails without it due to this https://github.com/etcd-io/etcd/issues/12124

Adding a comment

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:214]: Fixed

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:241]: The validator in `pkg/api/norman/customization/cluster/validator.go` ensures there is at least one node pool.

File: pkg/controllers/management/clusterprovisioner/provisioner.go
Comment (cmurphy) [reply pkg/controllers/management/clusterprovisioner/provisioner.go:122]: Added a `skipOperatorCluster` function for this.

**Comment by cmurphy** [comment]
> We should probably create a package contaaining logic that is used in multiple controllers such as deployOperator, setUnknown, setTrue, setFalse, aand generateValuesYaml.

Added a commit to do this.

File: pkg/systemtemplate/import.go
Comment (thedadams) [inline pkg/systemtemplate/import.go:78]: A `nil` array has length 0.
```suggestion
	if cluster != nil && len(cluster.Spec.AgentEnvVars) > 0 {
```

File: pkg/api/norman/customization/cluster/validator.go
Comment (thedadams) [inline pkg/api/norman/customization/cluster/validator.go:675]: There is already a name-collision validation check for all clusters. A user cannot have two clusters with the exact same name: https://github.com/rancher/rancher/blob/9a3562666ec2022358153d2c7a9193964ec57db3/pkg/api/norman/store/cluster/cluster_store.go#L613

File: pkg/controllers/management/clusterdeploy/clusterdeploy.go
Comment (thedadams) [reply pkg/controllers/management/clusterdeploy/clusterdeploy.go:129]: The way this works in the UI now is the following:
1. A user creates a private cluster.
2. Once the cluster is running, the UI gives the user the option to see the import command.
3. If the user clicks to see the import command, the UI makes a GET request to `/v3/clusterregistrationtokens` to see if the cluster has any registration tokens.
3a. If the cluster does not have any registration tokens, the UI makes a POST request to `/v3/clusterregistrationtokens` to create a token and the repeats step (3).

You are essentially doing (3) and (3a) automatically for every cluster that is created in Rancher. This is not the correct flow because not every cluster needs a registration token.

I believe removing this code and letting the UI handle this as above is the correct way to do it. I'm happy to help if you would like to test this with your development environment.

File: pkg/dialer/factory.go
Comment (thedadams) [inline pkg/dialer/factory.go:83]:
```suggestion
			!cluster.Status.GKEStatus.UpstreamSpec.PrivateClusterConfig.EnablePrivateEndpoint {
```

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (thedadams) [inline pkg/controllers/management/gke/gke_cluster_handler.go:88]: nit:
```suggestion
	if cluster == nil || cluster.DeletionTimestamp != nil || cluster.Spec.GKEConfig == nil {
		return cluster, nil
	}
```

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (thedadams) [inline pkg/controllers/management/gke/gke_cluster_handler.go:308]: It looks like this function could have a more generic name and moved to `clusteroperator/utils.go`. This would also require the EKS version to be deleted in favor of the more generic version.

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (thedadams) [inline pkg/controllers/management/gke/gke_cluster_handler.go:350]: It also looks like this function could be moved to `clusteroperator/utils.go`.

File: pkg/controllers/management/clusteroperator/utils.go
Comment (thedadams) [inline pkg/controllers/management/clusteroperator/utils.go:1]: This is excellent!

**Review by thedadams — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:301]: This one is a little tricky because it invokes `e.recordAppliedSpec` which is not generic.

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go]: done

**Review by thedadams — DISMISSED** [review]

File: pkg/api/norman/customization/cluster/validator.go
Comment (rmweir) [reply pkg/api/norman/customization/cluster/validator.go:565]: In the EKS example nothing is happening if ok is false. In your code an error is being returned. I think I was incorrect about the non-import part, both need it but only on a POST request. It is important to consider that when a user is just updating a single field, that is the only field that will be sent instead of sending the entire cluster/config.

Comment (cmurphy) [reply pkg/api/norman/customization/cluster/validator.go]: Got it, fixed by moving it down

**Review by thedadams — APPROVED** [review]

File: pkg/api/norman/customization/cluster/validator.go
Comment (rmweir) [reply pkg/api/norman/customization/cluster/validator.go]: The difference is that in EKS if ok is false, then it is not checked. I was incorrect on the import part. We need to validate POST requests, be them for an import or non import cluster, or any PUT requests that are changing the actual field. It is important to note that when updating something in the UI that only the field(s) that are being changed may be sent.

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (rmweir) [inline pkg/controllers/management/gke/gke_cluster_handler.go:152]: Did you test if this works for gke? Like if you add a cloud credential, then delete it in gke, does this catch that?

**Review by rmweir — CHANGES_REQUESTED** [review]
LGTM on everything non-syncing related besides 1 comment in this review and separate comment about credential validation. @mjura is nearly finished with syncing and it has changed quite a bit since being included in here.

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:152]: I had not tested it. It does not work, but not because of the message string: https://github.com/rancher/gke-operator/pull/17

File: pkg/api/norman/customization/cluster/validator.go
Comment (cmurphy) [reply pkg/api/norman/customization/cluster/validator.go]: Changed this to look like the EKS validator, it should now activate on either POST or PUT but only if the field is actually present in the config.

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (rmweir) [reply pkg/controllers/management/gke/gke_cluster_handler.go:152]: Okay, once that other PR is merged it is something that should be checked.

**Review by rmweir — APPROVED** [review]
LGTM, made comment about checking error for auth errors. Can either wait for related PR to be merged and test it out before merging this one or just merge this PR and check later.

**Review by thedadams — DISMISSED** [review]

File: pkg/controllers/management/clusteroperator/utils.go
Comment (cbron) [inline pkg/controllers/management/clusteroperator/utils.go:66]: We can use `localCluster` const here

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cbron) [reply pkg/controllers/management/gke/gke_cluster_handler.go:152]: That other PR is ready to merge, does that mean we can test this now ?

**Review by cbron — COMMENTED** [review]
minor things, otherwise lgtm, but need to wait for refresher PR

File: pkg/controllers/management/clusteroperator/utils.go
Comment (cmurphy) [reply pkg/controllers/management/clusteroperator/utils.go:66]: done

File: pkg/controllers/management/gke/gke_cluster_handler.go
Comment (cmurphy) [reply pkg/controllers/management/gke/gke_cluster_handler.go:152]: Sorry, I should have been more clear. I *had* not tested it but I have now. With the other pull request, it works - there is a string `403` in the error message that can be matched against.

**Review by cbron — DISMISSED** [review]

**Comment by cbron** [comment]
Needs to be rebased

**Review by cbron — APPROVED** [review]

**Comment by cmurphy** [comment]
Backport: https://github.com/rancher/rancher/pull/31907

**Review by cbron — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**
